src/audio/volume.c

```diff
@@ -250,6 +264,61 @@ static inline void volume_set_chan(struct comp_dev *dev, int chan, uint32_t vol)
 		v = VOL_MAX;
 
 	cd->tvolume[chan] = v;
+
+	/* Check ramp type */
+	switch (pga->ramp) {
+	case SOF_VOLUME_LINEAR:
+
+		/* Assume the ramp length (initial_ramp [ms]) describes
+		 * time of mute to 0 dB ramp. The actual volume scale min/max
+		 * is not known. If initial_ramp is zero (or negative) use
+		 * step size that reaches 0 dB gain in one step (VOL_ZERO_DB).
+		 */
+		if (pga->initial_ramp > 0)
+			inc = MAX(VOL_RAMP_STEP_CONST / pga->initial_ramp, 1);
+		else
+			inc = VOL_ZERO_DB;
+
+		delta = cd->tvolume[chan] - cd->volume[chan];
+		if (delta < 0)
+			delta_abs = -delta;
+		else
+			delta_abs = delta;
+
+		trace_volume("chan = %d", chan);
+		trace_volume("delta = %d", delta);
+		trace_volume("delta_abs = %d", delta_abs);
+
+		/* If the ramp completion would take longer than initial_ramp
+		 * increase the step size to make this ramp complete at
+		 * initial_ramp time. Calculate estimated ramp length [us].
+		 */
+		t_us = delta_abs * VOL_RAMP_UPDATE_US / inc;
+		if (t_us > pga->initial_ramp * 1000) {
+			inc = (int32_t)((int64_t)inc * delta_abs / VOL_ZERO_DB);
+			inc = MAX(inc, 1);
+			t_us = delta_abs * VOL_RAMP_UPDATE_US / inc;
+		}
+
+		/* Invert sign for volume down ramp step */
+		if (delta < 0)
+			inc = -inc;
+
+		cd->ramp_increment[chan] = inc;
+
+		trace_volume("ramp_increment = %d", cd->ramp_increment[chan]);
```

Comment:
Can be done in one trace message.

Reply:
Also these are not needed. The t_us recompute is just for trace and can be removed also.

Reply:
This could be tracev().